### ISSUE TITLE
Inconsistent behavior of F.conv2d(...,padding) and F.pad

### ISSUE THREAD

**Issue by codinfox** [pytorch/pytorch#8019]
## Issue description

I am getting different results from `F.conv2d(feat, kern, padding=(1,1))` and `F.pad(feat, (1,1,1,1)); F.conv2d(feat, kern)`.

What's more interesting is in the example I provided below, although both arguments for `F.conv2d` are non-negative, I still get some negative numbers as results.

## Code example
Download the data from here: https://drive.google.com/open?id=1Drgz7bTjLhPtpscvZK2daRx8Adqo5-b9

```python
In [2]: params = torch.load('params.pt')

In [3]: import torch.nn.functional as F

In [4]: a = F.conv2d(params['feat'], params['kern'], padding=(1,1))

In [5]: b = F.pad(params['feat'], (1,1,1,1))

In [6]: b = F.conv2d(b, params['kern'])

In [7]: (a != b).sum()
Out[7]: tensor(1.0868e+05, device='cuda:0')

In [8]: (a < 0).sum()
Out[8]: tensor(2531, device='cuda:0')

In [9]: (b < 0).sum()
Out[9]: tensor(0, device='cuda:0')

In [10]: (params['kern'] < 0).sum()
Out[10]: tensor(0, device='cuda:0')

In [11]: (params['feat'] < 0).sum()
Out[11]: tensor(0, device='cuda:0')

In [12]: a.shape
Out[12]: torch.Size([16, 512, 4, 4])

In [13]: b.shape
Out[13]: torch.Size([16, 512, 4, 4])
```

## System Info
Collecting environment information...
PyTorch version: 0.4.0
Is debug build: No
CUDA used to build PyTorch: 8.0.61

OS: Ubuntu 14.04.5 LTS
GCC version: (Ubuntu 4.8.5-4ubuntu8~14.04.2) 4.8.5
CMake version: version 3.2.2

Python version: 3.5
Is CUDA available: Yes
CUDA runtime version: 8.0.61
GPU models and configuration: GPU 0: GeForce GTX 1080
Nvidia driver version: 384.111
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.7.0.5
/usr/local/cuda-8.0/lib64/libcudnn.so
/usr/local/cuda-8.0/lib64/libcudnn.so.5
/usr/local/cuda-8.0/lib64/libcudnn.so.5.1.10
/usr/local/cuda-8.0/lib64/libcudnn.so.7
/usr/local/cuda-8.0/lib64/libcudnn.so.7.0.5
/usr/local/cuda-8.0/lib64/libcudnn_static.a

Versions of relevant libraries:
[pip] numpy (1.13.3)
[pip] numpydoc (0.7.0)
[pip] torch (0.4.0)
[pip] torchvision (0.2.0)
[conda] pytorch 0.4.0 py35_cuda8.0.61_cudnn7.1.2_1 pytorch
[conda] torchvision 0.2.0 py35_0

**Comment by SsnL** [comment]
This appears to be related cudnn. I can only reproduce this with cudnn enabled and nondeterministic algorithm used. With `torch.backends.cudnn.deterministic=False`, the problem goes away.

**Comment by codinfox** [comment]
Got it. I think you mean `torch.backends.cudnn.deterministic=True`

**Comment by codinfox** [comment]
Hmm. Actually, can this be documented somewhere? Also, according to the [cudnn doc](https://docs.nvidia.com/deeplearning/sdk/cudnn-developer-guide/index.html#reproducibility), the non-determinism happens during backward pass. However in my case it happens in the forward pass. Any thoughts?
